lib/job.js

```diff
@@ -198,22 +198,29 @@ Job.prototype.promote = function(){
     });
 };
 
-Job.prototype.retry = function(){
-  var key = this.queue.toKey('wait');
-  var failed = this.queue.toKey('failed');
-  var channel = this.queue.toKey('jobs');
-  var multi = this.queue.multi();
+Job.prototype.retry = function() {
   var _this = this;
-
-  multi.srem(failed, this.jobId);
-  // if queue is LIFO use rpushAsync
-  multi[(this.opts.lifo ? 'r' : 'l') + 'push'](key, this.jobId);
-  multi.publish(channel, this.jobId);
-
-  return multi.execAsync().then(function(){
-    _this.queue.emit('waiting', _this);
-    return _this;
+  var push = (this.opts.lifo ? 'R' : 'L') + 'PUSH';
+  var script = [
+    'if (redis.call("SREM", KEYS[2], ARGV[1]) == 1) or (redis.call("SREM", KEYS[3], ARGV[1]) == 1) then',
```

Comment:
I have mixed feelings about it. On the one hand, checking if it the job `EXISTS` is a nice way to gracefully handle leaky job ids. On the other hand, doing so can make errors more subtle and harder to track down, leaking job ids is an issue by itself. Hard fails are often easier to diagnose.